File: leetcode/problems/problem-specific/45.jump-game-ii/45.jump-game-ii.go
package main

func jump(nums []int) int {
	if len(nums) <= 0 {
		return 0
	}

	minJumps, curEnd, curFarthest := 0, 0, 0
	for i := 0; i < len(nums)-1; i++ {
		curFarthest = findMax(curFarthest, i+nums[i])
		if i == curEnd {
			curEnd = curFarthest
			minJumps++
		}
	}

	return minJumps
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

File: leetcode/problems/sessions/2025-10-25_2025-??-??/45.jump-game-ii.py
#
# @lc app=leetcode id=45 lang=python3
#
# [45] Jump Game II
#

# @lc code=start

# * Greedy Solution | O(n) Time | O(1) Space
# * n -> The length of nums array


class Solution:
    def jump(self, nums: List[int]) -> int:
        farthest_reached_idx = end_idx = min_jumps = 0
        for idx, num in enumerate(nums):
            if end_idx >= len(nums) - 1:
                break

            farthest_reached_idx = max(farthest_reached_idx, idx + num)
            if idx == end_idx:
                end_idx = farthest_reached_idx
                min_jumps += 1

        return min_jumps


# @lc code=end
